src/cpuid/src/common.rs

```diff
@@ -81,197 +70,137 @@ pub fn get_vendor_id_from_host() -> Result<[u8; 12], Error> {
-        // we can't tell if the CPUs are identical.
-        return false;
+/// Returns MSRs to be saved based on CPUID features that are enabled.
+///
+/// # Errors
+///
+/// When CPUID leaf 0 is not supported.
+#[cfg(cpuid)]
+pub fn msrs_to_save_by_cpuid(
+    cpuid: &kvm_bindings::CpuId,
+) -> Result<std::collections::HashSet<u32>, Leaf0NotFoundInCpuid> {
+    let vendor_id = get_vendor_id_from_cpuid(cpuid)?;
+    match &vendor_id {
+        VENDOR_ID_INTEL => Ok(intel_msrs_to_save_by_cpuid(cpuid)),
+        // We don't have MSR-CPUID dependencies set for other vendors yet.
+        _ => Ok(std::collections::HashSet::new()),
     }
-
-    // Try to get the feature information leaf from the host CPUID.
-    let host_feature_info_leaf = get_cpuid(leaf_0x1::LEAF_NUM, 0);
-
-    // The relevant information for this comparison is in the EAX register.
-    let host_feature_info_leaf_eax = match host_feature_info_leaf {
-        Ok(leaf) => leaf.eax,
-        Err(_) => {
-            // If this fails, we can't tell if the CPUs are identical.
-            return false;
-        }
-    };
-
-    // Search for the entry for leaf0x1.
-    let feature_info_leaf = cpuid
-        .as_slice()
-        .iter()
-        .find(|entry| entry.function == leaf_0x1::LEAF_NUM);
-
-    // The relevant information is in EAX.
-    let feature_info_leaf_eax = match feature_info_leaf {
-        Some(leaf) => leaf.eax,
-        None => {
-            // Fail fast if we can't retrieve the relevant
-            // information from CPUID.
-            return false;
-        }
-    };
-
-    // Validate that all of these properties are the same.
-    for elem in &[
-        leaf_0x1::eax::EXTENDED_FAMILY_ID_BITRANGE,
-        leaf_0x1::eax::EXTENDED_PROCESSOR_MODEL_BITRANGE,
-        leaf_0x1::eax::PROCESSOR_FAMILY_BITRANGE,
-        leaf_0x1::eax::PROCESSOR_MODEL_BITRANGE,
-        leaf_0x1::eax::STEPPING_BITRANGE,
-    ] {
-        if feature_info_leaf_eax.read_bits_in_range(elem)
-            != host_feature_info_leaf_eax.read_bits_in_range(elem)
-        {
-            return false;
-        }
-    }
-
-    true
 }
 
-/// Scans through the CPUID and determines if a feature bit is set.
-// TODO: This currently involves a linear search which would be improved
-//       when we'll refactor the cpuid crate.
-#[macro_export]
-macro_rules! cpuid_is_feature_set {
-    ($cpuid:ident, $leaf:expr, $index:expr, $reg:tt, $feature_bit:expr) => {{
-        let mut res = false;
-        for entry in $cpuid.as_slice().iter() {
-            if entry.function == $leaf && entry.index == $index {
-                if entry.$reg & (1 << $feature_bit) != 0 {
-                    res = true;
-                    break;
+/// Returns MSRs to be saved based on the Intel CPUID features that are enabled.
+#[cfg(cpuid)]
+#[must_use]
+pub fn intel_msrs_to_save_by_cpuid(cpuid: &kvm_bindings::CpuId) -> std::collections::HashSet<u32> {
```

Comment:
This has not been changed, it is https://github.com/firecracker-microvm/firecracker/blob/main/src/cpuid/src/template/intel/mod.rs#L31:L88 renamed with the `intel_` prefix since it is only applied to intel.